File: s04_hooks/hooks.py
from collections.abc import Callable
from typing import Any, Literal

from loguru import logger
from rich import print

from s04_hooks.constants import WORKDIR
from s04_hooks.permission import permission_hook

type HookEventType = Literal["UserPromptSubmit", "PreToolUse", "PostToolUse", "Stop"]
type HookCallback = Callable[..., str | None]

HOOKS: dict[HookEventType, list[HookCallback]] = {
    "UserPromptSubmit": [],
    "PreToolUse": [],
    "PostToolUse": [],
    "Stop": [],
}


def register_hook(event: HookEventType, callback: HookCallback) -> None:
    HOOKS[event].append(callback)


def trigger_hooks(event: HookEventType, *args: Any) -> str | None:
    for callback in HOOKS[event]:
        result = callback(*args)
        if result is not None:  # 返回值 != None → hook 说"停"
            return result
    return None


# --------------------- DEFAULT HOOKS ----------------------


def _context_inject_hook(query: str) -> str | None:
    """Inject current working directory info into every prompt."""
    logger.debug("[HOOK] UserPromptSubmit: working in {}", WORKDIR)
    return None  # return None = no modification, let prompt through


def _log_hook(block):
    """PreToolUse: log every tool call."""
    args_preview = str(list(block.input.values())[:2])[:60]
    print(f"[dim] [HOOK] use tool {block.name}({args_preview})[/dim]")
    return None


def _large_output_hook(block, output):
    """PostToolUse: warn on large output."""
    if len(str(output)) > 100000:
        print(
            f"[yellow] [HOOK] ⚠ Large output from {block.name}: {len(str(output))} chars[/yellow]"
        )
    return None


# Stop hook: print summary when loop is about to exit
def _summary_hook(messages: list):
    tool_count = sum(
        1
        for m in messages
        for b in (m.get("content") if isinstance(m.get("content"), list) else [])
        if isinstance(b, dict) and b.get("type") == "tool_result"
    )
    print(f"[dim] [HOOK] Stop: session used {tool_count} tool calls[/dim]")
    return None


register_hook("UserPromptSubmit", _context_inject_hook)
register_hook("PreToolUse", permission_hook)
register_hook("PreToolUse", _log_hook)
register_hook("PostToolUse", _large_output_hook)
register_hook("Stop", _summary_hook)

File: hooks.go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/anthropics/anthropic-sdk-go"
)

// HookEvent 事件类型
type HookEvent string

const (
	EventUserPromptSubmit HookEvent = "UserPromptSubmit"
	EventPreToolUse       HookEvent = "PreToolUse"
	EventPostToolUse      HookEvent = "PostToolUse"
	EventStop             HookEvent = "Stop"
)

// HookCallback 钩子回调函数，返回 nil 表示继续，返回 *string 表示中断
type HookCallback func(args ...any) *string

// hooks 全局钩子注册表
var hooks = map[HookEvent][]HookCallback{
	EventUserPromptSubmit: {},
	EventPreToolUse:       {},
	EventPostToolUse:      {},
	EventStop:             {},
}

// registerHook 注册钩子到指定事件
func registerHook(event HookEvent, cb HookCallback) {
	hooks[event] = append(hooks[event], cb)
}

// triggerHooks 触发事件的所有钩子，任一钩子返回非 nil 则中断并返回该值
func triggerHooks(event HookEvent, args ...any) *string {
	for _, cb := range hooks[event] {
		if result := cb(args...); result != nil {
			return result
		}
	}
	return nil
}

// ── 默认钩子 ──────────────────────────────────────────

// logHook 记录每次工具调用（PreToolUse）
func logHook(args ...any) *string {
	block, ok := args[0].(anthropic.ToolUseBlock)
	if !ok {
		return nil
	}
	inputJSON, _ := json.Marshal(block.Input)
	preview := truncate(string(inputJSON), 60)
	fmt.Printf("[HOOK] Used tool %s: (%s)\n", block.Name, preview)
	return nil
}

// largeOutputHook 警告过大的工具输出（PostToolUse）
func largeOutputHook(args ...any) *string {
	if len(args) < 2 {
		return nil
	}
	block, ok := args[0].(anthropic.ToolUseBlock)
	if !ok {
		return nil
	}
	output, ok := args[1].(string)
	if !ok {
		return nil
	}
	if len(output) > 100000 {
		fmt.Printf("%s [HOOK] ⚠ Large output from %s: %d chars\n", colorYellow("WARNING"), block.Name, len(output))
	}
	return nil
}

// summaryHook 在 agent 循环结束时打印工具调用统计（Stop）
func summaryHook(args ...any) *string {
	if len(args) < 1 {
		return nil
	}
	messages, ok := args[0].([]anthropic.MessageParam)
	if !ok {
		return nil
	}

	toolCount := 0
	for _, m := range messages {
		for _, block := range m.Content {
			if block.OfToolResult != nil {
				toolCount++
			}
		}
	}
	fmt.Printf("[HOOK] Stop: session used %d tool calls\n", toolCount)
	return nil
}

// initHooks 注册所有默认钩子
func initHooks() {
	registerHook(EventPreToolUse, permissionHook)
	registerHook(EventPreToolUse, logHook)
	registerHook(EventPostToolUse, largeOutputHook)
	registerHook(EventStop, summaryHook)
}
